### PR TITLE
add character zhangsong

### PR DESCRIPTION
**Pull request by Gaeric**
**强识**：出牌阶段开始时，你可以展示一名其他角色的一张手牌，然后当你于此阶段使用一张与展示牌类别相同的牌时，你可以摸一张牌。
**献图**：其他角色的出牌阶段开始时，你可以摸两张牌，然后交给其两张牌。若如此做，此阶段结束时，若其于此阶段内未杀死过角色，你失去1点体力。

### CODE REVIEW

File: src/core/skills/characters/yijiang2014/qiangzhi.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        toId: ownerId,
+        players: room.AlivePlayers.filter(
+          player => player.getCardIds(PlayerCardsArea.HandArea).length > 0 && player.Id !== ownerId,
+        ).map(player => player.Id),
+        requiredAmount: 1,
+        conversation: TranslationPack.translationJsonPatcher(
+          '{0}: please choose a player to show a hand card',
+          this.GeneralName,
+        ).extract(),
+      }),
```
Comment (donle) [inline src/core/skills/characters/yijiang2014/qiangzhi.ts:39]: add `triggeredBySkills`

File: src/core/skills/characters/yijiang2014/qiangzhi.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      ownerId,
+    );
+
+    const { selectedPlayers } = await room.onReceivingAsyncResponseFrom(
+      GameEventIdentifiers.AskForChoosingPlayerEvent,
+      ownerId,
+    );
+
+    if (selectedPlayers === undefined) {
+      return true;
```
Comment (donle) [inline src/core/skills/characters/yijiang2014/qiangzhi.ts:49]: return false

File: src/core/skills/characters/yijiang2014/qiangzhi.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      ).extract(),
+    });
+
+    room.getPlayerById(ownerId).setFlag(this.GeneralName, Sanguosha.getCardById(selectedCard).BaseType);
+
+    return true;
+  }
+}
+
+@ShadowSkill
```
Comment (donle) [inline src/core/skills/characters/yijiang2014/qiangzhi.ts:87]: clear flags in `whenRefresh` or `beforeUse`;

File: src/core/skills/characters/yijiang2014/xiantu.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    );
+
+    let { cardIds } = await room.onReceivingAsyncResponseFrom(GameEventIdentifiers.AskForSkillUseEvent, content.fromId);
+
+    if (cardIds === undefined) {
+      cardIds = room.getPlayerById(content.fromId).getCardIds().slice(0, 2);
+    }
+
+    // optimize hide card
+    room.moveCards({
```
Comment (donle) [inline src/core/skills/characters/yijiang2014/xiantu.ts:53]: await

**Review by donle — CHANGES_REQUESTED** [review]

File: src/core/skills/characters/yijiang2014/qiangzhi.ts (hunk 1 above)
Comment (Gaeric) [reply src/core/skills/characters/yijiang2014/qiangzhi.ts:39]: I don't understand what info you need, could you please give me a example?

File: src/core/skills/characters/yijiang2014/qiangzhi.ts (hunk 3 above)
Comment (Gaeric) [reply src/core/skills/characters/yijiang2014/qiangzhi.ts:101]: Refresh skill at PlayCardStage, and remove this flag when refresh skill.
Add contriant: Can't trigger skill in other player's phase.

File: src/core/skills/characters/yijiang2014/qiangzhi.ts (hunk 2 above)
Comment (Gaeric) [reply src/core/skills/characters/yijiang2014/qiangzhi.ts]: done

File: src/core/skills/characters/yijiang2014/qiangzhi.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    return stage === CardUseStage.CardUsing;
+  }
+
+  canUse(room: Room, owner: Player, event: ServerEventFinder<GameEventIdentifiers.CardUseEvent>) {
+    if (event.fromId !== owner.Id || room.CurrentPhasePlayer !== owner) {
+      return false;
+    }
+
+    return (
+      owner.getFlag(this.GeneralName) !== undefined &&
```
Comment (donle) [inline src/core/skills/characters/yijiang2014/qiangzhi.ts:116]: this line is not required since if it equals the flag type, it would not be undefined

**Review by donle — APPROVED** [review]

**Pull request merged**